modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/CalciteQueryProcessor.java

```diff
@@ -326,89 +327,114 @@ public PrepareServiceImpl prepareService() {
         QueryPlan plan = queryPlanCache().queryPlan(new CacheKey(schema.getName(), sql));
 
         if (plan != null) {
-            RootQuery<Object[]> qry = new RootQuery<>(
-                sql,
-                schema,
-                params,
-                qryCtx,
-                exchangeSvc,
-                (q) -> qryReg.unregister(q.id()),
-                log,
-                queryPlannerTimeout
+            return Collections.singletonList(
+                executeQuery(qryCtx, qry -> plan, schemaName, sql, null, params)
             );
+        }
 
-            qryReg.register(qry);
-
-            try {
-                return Collections.singletonList(executionSvc.executePlan(
-                    qry,
-                    plan
-                ));
-            }
-            catch (Exception e) {
-                boolean isCanceled = qry.isCancelled();
-
-                qry.cancel();
+        SqlNodeList qryList = Commons.parse(sql, FRAMEWORK_CONFIG.getParserConfig());
 
-                qryReg.unregister(qry.id());
+        List<FieldsQueryCursor<List<?>>> cursors = new ArrayList<>(qryList.size());
+        List<RootQuery<Object[]>> qrys = new ArrayList<>(qryList.size());
 
-                if (isCanceled)
-                    throw new IgniteSQLException("The query was cancelled while planning", IgniteQueryErrorCode.QUERY_CANCELED, e);
+        for (final SqlNode sqlNode: qryList) {
+            FieldsQueryCursor<List<?>> cursor = executeQuery(qryCtx, qry -> {
+                if (qryList.size() == 1) {
+                    return queryPlanCache().queryPlan(
+                        new CacheKey(schemaName, sql), // Use source SQL to avoid redundant parsing next time.
+                        () -> prepareSvc.prepareSingle(sqlNode, qry.planningContext())
+                    );
+                }
                 else
-                    throw e;
+                    return prepareSvc.prepareSingle(sqlNode, qry.planningContext());
+            }, schemaName, sqlNode.toString(), qrys, params);
 
-            }
+            cursors.add(cursor);
         }
 
-        SqlNodeList qryList = Commons.parse(sql, FRAMEWORK_CONFIG.getParserConfig());
-        List<FieldsQueryCursor<List<?>>> cursors = new ArrayList<>(qryList.size());
+        return cursors;
+    }
 
-        List<RootQuery<Object[]>> qrys = new ArrayList<>(qryList.size());
+    /** {@inheritDoc} */
+    @Override public List<FieldsQueryCursor<List<?>>> queryBatched(
+        @Nullable QueryContext qryCtx,
+        @Nullable String schemaName,
```

Comment:
`schemaName` is not `@Nullable` in parent method definition, and looks like it can't be null.